_examples/real-world-examples/server-sent-events/server/http.go

```diff
@@ -0,0 +1,318 @@
+	Content string `json:"content"`
+	Author  string `json:"author"`
+}
+
+func (router Router) CreatePost(w http.ResponseWriter, r *http.Request) {
+	req := CreatePostRequest{}
+	err := render.Decode(r, &req)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	post := NewPost(
+		watermill.NewUUID(),
+		req.Title,
+		req.Content,
+		req.Author,
+	)
+
+	err = router.addPost(r.Context(), post)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	w.WriteHeader(204)
+}
+
+func (router Router) GeneratePost(w http.ResponseWriter, r *http.Request) {
+	title := gofakeit.Sentence(5)
+	content := gofakeit.Sentence(20)
+	author := gofakeit.Name()
+
+	tagsCount := rand.Intn(3) + 1
+	for i := 0; i < tagsCount; i++ {
+		content += fmt.Sprintf(" #%v", generatedTags[rand.Intn(len(generatedTags))])
+	}
+
+	post := NewPost(
+		watermill.NewUUID(),
+		title,
+		content,
+		author,
+	)
+
+	err := router.addPost(r.Context(), post)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	w.WriteHeader(204)
+}
+
+func (router Router) addPost(ctx context.Context, post Post) error {
+	err := router.PostsStorage.Add(ctx, post)
+	if err != nil {
+		return err
+	}
+
+	event := PostCreated{
+		Post:       post,
+		OccurredAt: time.Now().UTC(),
+	}
+
+	err = router.Publisher.Publish(PostCreatedTopic, event)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type UpdatePostRequest struct {
+	ID      string `json:"id"`
+	Title   string `json:"title"`
+	Content string `json:"content"`
+}
+
+func (router Router) UpdatePost(w http.ResponseWriter, r *http.Request) {
+	req := UpdatePostRequest{}
+	err := render.Decode(r, &req)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	post, err := router.PostsStorage.ByID(r.Context(), req.ID)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	newPost := NewPost(
+		post.ID,
+		req.Title,
+		req.Content,
+		post.Author,
+	)
+
+	err = router.PostsStorage.Update(r.Context(), newPost)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	event := PostUpdated{
+		OriginalPost: post,
+		NewPost:      newPost,
+		OccurredAt:   time.Now().UTC(),
+	}
+
+	err = router.Publisher.Publish(PostUpdatedTopic, event)
+	if err != nil {
+		logAndWriteError(router.Logger, w, err)
+		return
+	}
+
+	w.WriteHeader(204)
+}
+
+type feedStreamAdapter struct {
+	storage FeedsStorage
+	logger  watermill.LoggerAdapter
+}
+
+func (f feedStreamAdapter) GetResponse(w http.ResponseWriter, r *http.Request) (response interface{}, ok bool) {
+	feedName := chi.URLParam(r, "name")
+
+	feed, err := f.storage.ByName(r.Context(), feedName)
+	if err != nil {
+		logAndWriteError(f.logger, w, err)
+		return nil, false
+	}
+
+	return feed, true
+}
+
+func (f feedStreamAdapter) Validate(r *http.Request, msg *message.Message) (ok bool) {
+	feedUpdated := FeedUpdated{}
+
+	err := json.Unmarshal(msg.Payload, &feedUpdated)
```

Comment:
it just came to my mind that metadata may be nice here as well (you don't need to unmarshal) but I think that this way is a bit more obvious

Reply:
It would be simpler from the code execution point of view, but I think it's easier to understand "real" events being published.